content/blog/explicit-apis-vs-magic-directives.mdx

```diff
@@ -0,0 +1,400 @@
+
+If you’re **prototyping**, you might care more about:
+
+- **Speed to first workflow.** Inngest’s event model may be easier to iterate on, but directives are faster to learn for someone new to this work. Add `"use workflow"` and your function is durable. No conceptual overhead.
+- **Simpler mental model for basic workflows.** For straightforward sequential tasks without coordination needs, WDK's approach means there’s less critical thinking to do.
+- **Keep everything with one provider.** You might already use Vercel for a lot. While Inngest works seamlessly with Vercel, you may still prefer consolidating your vendors.
+
+In short, if you're prototyping or building internal tools on Vercel, WDK might be faster for you to stand up. If you're scaling to enterprise with multiple tenants and complex coordination requirements, Inngest's architecture handles problems WDK doesn't address yet.
+
+## Understanding an Important Architectural Difference
+
+What exactly makes Inngest’s approach more suitable for removing complexity in production workflows? Some of it comes down to an early decision to switch from a directive model (Vercel’s approach) to explicit function calls.
+
+### The Directive Approach
+
+With directives, you write code that looks like a regular function:
+
+```tsx
+export async function processOrder(orderId: string) {
+  "use workflow";
+  
+  const order = await validateOrder(orderId);
+  const payment = await processPayment(order);
+  return await fulfillOrder(payment);
+}
+```
+
+When you write `"use workflow"` at the top of a function, something happens between your keyboard and production. Your code gets transformed, compiled into separate routes, and restructured in ways you don’t explicitly define. What you write isn't what runs—there's a compilation step that restructures your function into a durable workflow with hidden orchestration logic.
+
+### The Explicit Approach
+
+Inngest uses explicit function calls. What you write is what runs:
+
+```tsx
+export default inngest.createFunction(
+  { id: "process-order" },
+  { event: "order/created" },
+  async ({ event, step }) => {
+    const order = await step.run("validate", () => validateOrder(orderId));
+    const payment = await step.run("payment", () => processPayment(order));
+    return await step.run("fulfill", () => fulfillOrder(payment));
+  }
+);
+```
+
+There's no transformation layer—`step.run()` is a function call to Inngest's execution engine. Each step has an explicit ID that becomes its state key. This means you can refactor code, add logging, even rewrite in a different language, and running workflows continue correctly as long as step IDs stay the same.
+
+Functions react to events rather than direct calls. Send an `order/created` event and this function runs. Multiple functions can react to the same event, enabling natural fan-out without tight coupling.
```

Comment:
```suggestion
Functions react to events rather than direct calls. Send an `order/created` event and this function runs. Multiple functions can react to the same event, enabling natural [fan-out](/docs/guides/fan-out-jobs) without tight coupling.
```